tests/permissions/permissions_arches_default_deny_tests.py

```diff
@@ -0,0 +1,118 @@
+        """
+
+        implicit_permission = self.framework.user_can_read_resource(
+            self.user, self.resource_instance_id
+        )
+        resource = ResourceInstance.objects.get(
+            resourceinstanceid=self.resource_instance_id
+        )
+        self.framework.assign_perm("change_resourceinstance", self.group, resource)
+        can_access_without_view_permission = self.framework.user_can_read_resource(
+            self.user, self.resource_instance_id
+        )
+        self.framework.assign_perm("view_resourceinstance", self.group, resource)
+        can_access_with_view_permission = self.framework.user_can_read_resource(
+            self.user, self.resource_instance_id
+        )
+        self.assertTrue(
+            implicit_permission is False
+            and can_access_without_view_permission is False
+            and can_access_with_view_permission is True
+        )
+
+    def test_user_has_resource_model_permissions(self):
+        """
+        Tests that a user cannot access an instance if they have no access to any nodegroup.
+
+        """
+
+        resource = ResourceInstance.objects.get(
+            resourceinstanceid=self.resource_instance_id
+        )
+        nodes = Node.objects.filter(graph_id=resource.graph_id)
+        for node in nodes:
+            if node.nodegroup:
+                self.framework.assign_perm(
+                    "no_access_to_nodegroup", self.group, node.nodegroup
+                )
+        hasperms = self.framework.user_has_resource_model_permissions(
+            self.user, ["models.read_nodegroup"], resource
+        )
+        self.assertTrue(hasperms is False)
+
+    def test_get_restricted_users(self):
+        """
+        Tests that users are properly identified as restricted.
+        """
+
+        resource = ResourceInstance.objects.get(
+            resourceinstanceid=self.resource_instance_id
+        )
+        self.framework.assign_perm(
+            "no_access_to_resourceinstance", self.group, resource
+        )
+        ben = self.user
+        jim = User.objects.get(username="jim")
+        sam = User.objects.get(username="sam")
+        admin = User.objects.get(username="admin")
+        self.framework.assign_perm("view_resourceinstance", ben, resource)
+        self.framework.assign_perm("change_resourceinstance", jim, resource)
+
+        restrictions = self.framework.get_restricted_users(resource)
+
+        results = [
+            jim.id in restrictions["cannot_read"],
+            ben.id in restrictions["cannot_write"],
+            sam.id in restrictions["cannot_delete"],
+            sam.id in restrictions["no_access"],
+            admin.id not in restrictions["cannot_read"],
+            admin.id not in restrictions["cannot_write"],
+            admin.id not in restrictions["cannot_delete"],
+            admin.id not in restrictions["no_access"],
+        ]
+
+        self.assertTrue(all(results) is True)
```

Comment:
Probably better as `with self.subTest(...`

Reply:
That's a neat trick.